File: frontend/src/api/harvest.ts
import axios from 'axios'

const api = axios.create({ baseURL: 'http://localhost:8000' })

// ── request / response types ───────────────────────────────────────────

export interface HarvestEntry {
  field_id:          number
  field_name:        number
  picker_id:         number
  box_number:        number
  box_type_id:       number
  harvest_date:      string
  scan_date:         string
  picker_first_name: string
  picker_last_name:  string
  box_name:          string
  box_net_weight_kg: number
}

export interface BarcodeCheckResponse {
  barcode:   string
  valid:     boolean
  reason:    string | null
  scan_date: string | null
}

export interface BulkScanRequest {
  field_id:     number
  box_type_id:  number
  harvest_date: string
  barcodes:     string[]
}

export interface BulkScanResult {
  success:  boolean
  accepted: HarvestEntry[]
  problems: BarcodeCheckResponse[]
}

export interface PaginatedEntries {
  items:     HarvestEntry[]
  total:     number
  page:      number
  page_size: number
  pages:     number
}

// ── stats types ────────────────────────────────────────────────────────

export interface DailyStatEntry {
  harvest_date: string
  box_type_id:  number
  count:        number
}

export interface DailyStatsResponse {
  stats: DailyStatEntry[]
  total: number
}

export interface HarvestOverview {
  total_pickers: number
  total_scanned: number
  total_kg:      number
}

export interface PickerStat {
  picker_id:    number
  first_name:   string
  last_name:    string
  origin_place: string | null
  total_boxes:  number
  total_kg:     number
}

export interface DayBoxBreakdown {
  kg:        number
  box_types: Record<string, {
    count:         number
    net_weight_kg: number
    total_kg:      number
  }>
}

export interface PickerBoxStat {
  picker_id:       number
  first_name:      string
  last_name:       string
  national_id:     string
  phone:           string | null
  bank_info:       string | null
  origin_place:    string | null
  total_kg:        number
  total_boxes:     number
  total_box_types: Record<string, number>
  days:            Record<string, DayBoxBreakdown>
}

export interface FieldStat {
  field_id:    number
  field_name:  string
  description: string | null
  total_boxes: number
  total_kg:    number
}

// ── export types ────────────────────────────────────────────────────────

export interface PickerDetailExportEntry {
  barcode:       string
  box_name:      string
  net_weight_kg: number
  field_name:    string
  harvest_date:  string
}

export interface PickerDetailBoxSummary {
  box_name:      string
  net_weight_kg: number
  count:         number
}

export interface PickerDetailExportRow {
  picker_id:        number
  first_name:       string
  last_name:        string
  national_id:      string
  origin_place:     string | null
  phone:            string
  total_boxes:      number
  total_kg:         number
  box_type_summary: PickerDetailBoxSummary[]
  entries:          PickerDetailExportEntry[]
}

// ── helpers ────────────────────────────────────────────────────────────

function dateParams(from?: string, to?: string): URLSearchParams {
  const p = new URLSearchParams()
  if (from) p.append('from_date', from)
  if (to)   p.append('to_date',   to)
  return p
}

// ── scan endpoints ─────────────────────────────────────────────────────

export const checkBarcode = (barcode: string) =>
  api.post<BarcodeCheckResponse>('/harvest/check', { barcode }).then(r => r.data)

export const bulkScan = (data: BulkScanRequest) =>
  api.post<BulkScanResult>('/harvest/commit', data).then(r => r.data)

// ── entry listing ──────────────────────────────────────────────────────

export const getEntries = (page = 1, pageSize = 25, search = '') => {
  const params = new URLSearchParams()
  params.append('page',      String(page))
  params.append('page_size', String(pageSize))
  if (search) params.append('search', search)
  return api.get<PaginatedEntries>(`/harvest/entries?${params}`).then(r => r.data)
}

// ── stats endpoints ────────────────────────────────────────────────────

export const getHarvestOverview = () =>
  api.get<HarvestOverview>('/harvest/stats/overview').then(r => r.data)

export const getDailyStats = (from?: string, to?: string) =>
  api.get<DailyStatsResponse>(`/harvest/stats/daily?${dateParams(from, to)}`).then(r => r.data)

export const getPickerStats = () =>
  api.get<PickerStat[]>('/harvest/stats/pickers').then(r => r.data)

export const getPickerBoxStats = (from?: string, to?: string) =>
  api.get<PickerBoxStat[]>(`/harvest/stats/pickers/boxes?${dateParams(from, to)}`).then(r => r.data)

export const getFieldStats = (from?: string, to?: string) =>
  api.get<FieldStat[]>(`/harvest/stats/fields?${dateParams(from, to)}`).then(r => r.data)

// ── export endpoints ────────────────────────────────────────────────────

export const getPickerDetailExport = () =>
  api.get<PickerDetailExportRow[]>('/harvest/export/picker-detail').then(r => r.data)

File: backend/app/routers/harvest.py
from fastapi import APIRouter, Depends
from sqlalchemy import select
from sqlalchemy import func
from sqlalchemy.ext.asyncio import AsyncSession
from datetime import date
from app.database import get_db
from app.entities.harvest_entry import HarvestEntry
from app.entities.print_batch import PrintBatch
from app.entities.picker import Picker
from app.entities.box import Box
from app.schemas.harvest_entry import (
    BarcodeCheckRequest,
    BarcodeCheckResponse,
    BulkScanRequest,
    BulkScanResult,
    HarvestEntryResponse,
    DailyStatEntry,
    DailyStatsResponse
)

router = APIRouter(prefix="/harvest", tags=["harvest"])


def parse_barcode(barcode: str) -> tuple[int, int, int] | None:
    try:
        parts = barcode.strip().split("-")
        if len(parts) != 3:
            return None
        fruit_id   = int(parts[0])
        picker_id  = int(parts[1])
        box_number = int(parts[2])
        return fruit_id, picker_id, box_number
    except ValueError:
        return None


async def check_barcode(barcode: str, db: AsyncSession) -> BarcodeCheckResponse:
    parsed = parse_barcode(barcode)
    if not parsed:
        return BarcodeCheckResponse(
            barcode=barcode,
            valid=False,
            reason="invalid_format"
        )

    fruit_id, picker_id, box_number = parsed

    existing = await db.get(HarvestEntry, (picker_id, box_number))
    if existing:
        return BarcodeCheckResponse(
            barcode=barcode,
            valid=False,
            reason="already_scanned",
            scan_date=existing.scan_date
        )

    result = await db.execute(
        select(PrintBatch).where(
            PrintBatch.picker_id == picker_id,
            PrintBatch.box_number_from <= box_number,
            PrintBatch.box_number_to >= box_number
        )
    )
    if result.scalar_one_or_none() is None:
        return BarcodeCheckResponse(
            barcode=barcode,
            valid=False,
            reason="never_printed"
        )

    return BarcodeCheckResponse(barcode=barcode, valid=True)


@router.post("/check", response_model=BarcodeCheckResponse)
async def check_single_barcode(
    data: BarcodeCheckRequest,
    db: AsyncSession = Depends(get_db)
):
    return await check_barcode(data.barcode, db)


@router.post("/scan", response_model=BulkScanResult)
async def bulk_scan(
    data: BulkScanRequest,
    db: AsyncSession = Depends(get_db)
):
    problems = []
    entries  = []

    for barcode in data.barcodes:
        result = await check_barcode(barcode, db)
        if not result.valid:
            problems.append(result)
        else:
            parsed = parse_barcode(barcode)
            fruit_id, picker_id, box_number = parsed
            entries.append(HarvestEntry(
                fruit_id=fruit_id,
                picker_id=picker_id,
                box_number=box_number,
                box_type_id=data.box_type_id,
                harvest_date=data.harvest_date,
            ))

    if problems:
        return BulkScanResult(success=False, accepted=[], problems=problems)

    for entry in entries:
        db.add(entry)
    await db.commit()
    for entry in entries:
        await db.refresh(entry)

    return BulkScanResult(
        success=True,
        accepted=[HarvestEntryResponse.model_validate(e) for e in entries],
        problems=[]
    )


@router.get("/", response_model=list[HarvestEntryResponse])
async def get_all_entries(db: AsyncSession = Depends(get_db)):
    result = await db.execute(
        select(HarvestEntry).order_by(HarvestEntry.scan_date.desc())
    )
    return result.scalars().all()


@router.get("/stats", response_model=DailyStatsResponse)
async def get_daily_stats(
    from_date: date | None = None,
    to_date:   date | None = None,
    db: AsyncSession = Depends(get_db)
):
    query = (
        select(
            HarvestEntry.harvest_date,
            HarvestEntry.box_type_id,
            func.count().label("count")
        )
        .group_by(HarvestEntry.harvest_date, HarvestEntry.box_type_id)
        .order_by(HarvestEntry.harvest_date)
    )

    if from_date:
        query = query.where(HarvestEntry.harvest_date >= from_date)
    if to_date:
        query = query.where(HarvestEntry.harvest_date <= to_date)

    result = await db.execute(query)
    rows   = result.all()

    stats = [
        DailyStatEntry(
            harvest_date=row.harvest_date,
            box_type_id=row.box_type_id,
            count=row.count
        )
        for row in rows
    ]

    return DailyStatsResponse(stats=stats, total=sum(s.count for s in stats))


@router.get("/overview")
async def get_overview(db: AsyncSession = Depends(get_db)):
    # total pickers
    picker_result = await db.execute(select(func.count()).select_from(Picker))
    total_pickers = picker_result.scalar() or 0

    # total boxes scanned
    scan_result = await db.execute(select(func.count()).select_from(HarvestEntry))
    total_scanned = scan_result.scalar() or 0

    # total kg — join harvest_entry with box to get net_weight_kg per entry
    kg_result = await db.execute(
        select(func.sum(Box.net_weight_kg))
        .join(HarvestEntry, HarvestEntry.box_type_id == Box.box_id)
    )
    total_kg = float(kg_result.scalar() or 0)

    return {
        "total_pickers": total_pickers,
        "total_scanned": total_scanned,
        "total_kg":      round(total_kg, 3),
    }